### PR TITLE
feat: initial implementation 

### PR DESCRIPTION
**Pull request by CptSchnitz**

### CODE REVIEW

File: scripts/validate.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  if (files.every((file) => file.isDirectory())) {
+    directories.push(...files.map((file) => path.join(directory, file.name)));
+    continue;
+  }
+
+  let fileOrderCounter = 1;
+
+  for (const file of files) {
+    await asyncLocalStorage.run({ directory, file: file.name }, async () => {
+      // the correct structure is is either only directories or only files, so if its not a file at this stage, it is an error
```
Comment (shimoncohen) [inline scripts/validate.ts:161]: Remove duplicate `is`.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -1,2 +1,131 @@
...
+  }
+}
+```
+
+
+## Custom schema properties
+In order to enable extra abilities and features we have added custom properties that can be added to schemas.
+
+- `x-env-value` - The value from which to pull environment variables to override the config values.
+- `x-populate-as-env` - ***experimental*** Whether to place the value into the environment variable defined in `x-env-value`.
```
[the remark this answers is not in the document]
Comment (CptSchnitz) [reply README.md:100]: If the env is defined, nothing would happen.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,131 @@
...
+```
+
+## Validations
+The repo contains a validation script to make sure all the schemas are valid.
+To run the script write the following command into your terminal: `npm run validate`.
+
+The script contains the following validations:
+- Each folder contains either folders or only schema files (schema | markdown).
+- The schema files are in the correct format - Typescript or JSON.
+- The files in the folders are in ascending order (v1,v2,v3,...).
```
[the remark this answers is not in the document]
Comment (CptSchnitz) [reply README.md:122]: We do check there is no skipping, its to keep order and consistency.

File: schemas/common/db/partial/v1.schema.json [hunk 4]
```diff
@@ -0,0 +1,74 @@
+{
+  "$id": "https://mapcolonies.com/common/db/partial/v1",
+  "description": "database configuration for postgres",
+  "type": "object",
+  "required": ["ssl"],
```
[the remark this answers is not in the document]
Comment (CptSchnitz) [reply schemas/common/db/partial/v1.schema.json:5]: host and `port` have default values so there is no need for them to be required.
`ssl` is required as i want `enabled`  to always have a value.
In more general sense, for now the schemas are temporary until we release the package for general use.

File: schemas/common/db/partial/v1.schema.json [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        }
+      },
+      "then": {
+        "required": ["ca", "cert", "key"]
+      },
+      "else": {
+        "properties": {
+          "enabled": {
+            "const": false
+          }
```
[the remark this answers is not in the document]
Comment (CptSchnitz) [reply schemas/common/db/partial/v1.schema.json:69]: The logic is to set the value of `enabled` to `false` unless its `true`.

File: scripts/build.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  encoding: 'utf-8',
+});
+filesToDelete.push(symbolFilePath);
+
+// loop over all the files in the schemas directory and create the build files
+for await (const file of filesTreeGenerator(schemasFolder)) {
+  const directory = file.path;
+  const fullPath = path.join(directory, file.name);
+  const fileVersion = file.name.split('.')[0];
+  const schemaName = camelCase(directory.substring(7).replaceAll(path.sep, '_') + '_' + fileVersion);
```
[the remark this answers is not in the document]
Comment (CptSchnitz) [reply scripts/build.ts]: changed to not be a magic number

File: scripts/validate.ts (hunk 1 above)
Comment (CptSchnitz) [reply scripts/validate.ts]: removed

File: README.md [hunk 7]
```diff
@@ -1,2 +1,131 @@
...
+  "properties": {
+    "name": {
+      "type": "string",
+    }
+  }
+}
+```
+- The JSON Schema version used is [Draft-07](https://json-schema.org/draft-07/json-schema-release-notes).
+- Each schema should contain the ID field with the value based on the relative location of the schema in the schemas folder.
+- If the schema is in Typescript, it should have an default export that is a valid JSON Schema. For convenience, the package [typebox](https://github.com/sinclairzx81/typebox) is installed to help generating JSON Schemas.
```
[the remark this answers is not in the document]
Comment (CptSchnitz) [reply README.md]: removed

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply README.md:100]: Ok.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply README.md:122]: Ok, I don't see a problem with that currently.

File: schemas/common/db/partial/v1.schema.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply schemas/common/db/partial/v1.schema.json:5]: Ok, sounds good anyway.

File: schemas/common/db/partial/v1.schema.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply schemas/common/db/partial/v1.schema.json:69]: Sounds good!

File: scripts/build.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply scripts/build.ts]: Great :)

**Review by shimoncohen — APPROVED** [review]

**Review by syncush — APPROVED** [review]

**Pull request merged**